### PR TITLE
Consolidate tombstone handling

### PR DESCRIPTION
**Pull request by howardjohn**
Currently, we have this logic spread throughout the codebase, introducing boilerplate and risk of bugs. This unifies the implementation

### CODE REVIEW

File: pkg/kube/controllers/common.go [hunk 1]
```diff
@@ -179,23 +184,33 @@ func filteredObjectHandler(handler func(o Object), onlyIncludeSpecChanges bool,
 	}
 }
 
-func extractObject(obj any) Object {
-	o, ok := obj.(Object)
+// Extract pulls a T from obj, handling tombstones.
+// This will return nil if the object cannot be extracted.
+func Extract[T Object](obj any) T {
```
Comment (hzxuzhonghu) [inline pkg/kube/controllers/common.go:189]: We can not get the valid error info from the logs of this function, we need to print type as well

Comment (howardjohn) [reply pkg/kube/controllers/common.go:189]: 👍 will improve the message

Comment (howardjohn) [reply pkg/kube/controllers/common.go:189]: Added tests and now the log says: `tombstone contained object that is not an object (key:, obj:*v1.Pod)` (note: in real world key will normally be there, this is just from test that doesn't set it)

**Review by ericvn — APPROVED** [review]
LGTM - Thanks.

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**